extensions/amp-analytics/0.1/visibility-helper.js

```diff
@@ -0,0 +1,397 @@
+
+    /** @private {boolean} */
+    this.blocked_ = false;
+
+    /** @private {?number} */
+    this.scheduledRunId_ = null;
+
+    /** @private {boolean} */
+    this.matchesVisibility_ = false;
+
+    /** @private {time} */
+    this.continuousTime_ = 0;
+
+    /** @private {time} */
+    this.maxContinuousVisibleTime_ = 0;
+
+    /** @private {time} */
+    this.totalVisibleTime_ = 0;
+
+    /** @private {time} */
+    this.firstSeenTime_ = 0;
+
+    /** @private {time} */
+    this.lastSeenTime_ = 0;
+
+    /** @private {time} */
+    this.fistVisibleTime_ = 0;
+
+    /** @private {time} */
+    this.lastVisibleTime_ = 0;
+
+    /** @private {time} */
+    this.loadTimeVisibility_ = 0;
+
+    /** @private {number} */
+    this.minVisiblePercentage_ = 0;
+
+    /** @private {number} */
+    this.maxVisiblePercentage_ = 0;
+
+    /** @private {time} */
+    this.lastVisibleUpdateTime_ = 0;
+
+    if (this.parent_) {
+      this.parent_.addChild_(this);
+    }
+
+    if (this.getVisibility() > 0) {
+      this.update();
+    }
+  }
+
+  /** @override */
+  dispose() {
+    if (this.parent_) {
+      this.parent_.removeChild_(this);
+    }
+    if (this.scheduledRunId_) {
+      clearTimeout(this.scheduledRunId_);
+      this.scheduledRunId_ = null;
+    }
+    this.unsubscribe_.forEach(unsubscribe => {
+      unsubscribe();
+    });
+    this.unsubscribe_.length = 0;
+    this.eventResolver_ = null;
+  }
+
+  /**
+   * Adds the unsubscribe handler that will be called when this visibility
+   * helper is destroyed.
+   * @param {!UnsubscribeDef} handler
+   */
+  unsubscribe(handler) {
+    this.unsubscribe_.push(handler);
+  }
+
+  /**
+   * Adds the event handler that will be called when all visibility conditions
+   * have been met.
+   * @param {function()} handler
+   */
+  onEvent(handler) {
+    this.eventPromise_.then(handler);
+  }
+
+  /**
+   * Sets visibility of this object. See `getVisibility()` for the final
+   * visibility calculations.
+   * @param {number} visibility
+   */
+  setVisibility(visibility) {
+    this.ownVisibility_ = visibility;
+    this.update();
+  }
+
+  /**
+   * Sets whether this object is blocked. Blocking means that visibility is
+   * not ready to be calculated, e.g. because an element has not yet
+   * sufficiently rendered. See `getVisibility()` for the final
+   * visibility calculations.
+   * @param {boolean} blocked
+   */
+  setBlocked(blocked) {
+    this.blocked_ = blocked;
+    this.update();
+  }
+
+  /**
+   * Returns the final visibility. It depends on the following factors:
+   *  1. This object's visibility.
+   *  2. Whether the object is blocked.
+   *  3. The parent's visibility.
+   * @return {number}
+   */
+  getVisibility() {
+    const ownVisibility = this.ownVisibility_ * (this.blocked_ ? 0 : 1);
+    if (!this.parent_) {
+      return ownVisibility;
+    }
+    if (this.factorParent_) {
+      return ownVisibility * this.parent_.getVisibility();
```

Comment:
This is calling `#getVisibility` on `#parent` (`ownVisibility * this.parent_.getVisibility();`). Is this supposed to be `#factorParent`?